crates/bevy_ecs/src/bundle.rs

```diff
@@ -1346,11 +1349,39 @@ impl Bundles {
                 // - it was created in the same order as the components in T
                 unsafe { BundleInfo::new(core::any::type_name::<T>(), components, component_ids, id) };
             bundle_infos.push(bundle_info);
+
+            // Be sure to update component bundle with requires
+            contributed_bundle_ids.remove(&TypeId::of::<T>());
+
             id
         });
         id
     }
 
+    /// Registers a new [`BundleInfo`], which contains both explicit and required components for a statically known type.
+    ///
+    /// Also registers all the components in the bundle.
+    pub(crate) fn register_contributed_bundle_info<T: Bundle>(
+        &mut self,
+        components: &mut Components,
+        storages: &mut Storages,
+    ) -> BundleId {
+        if let Some(id) = self.contributed_bundle_ids.get(&TypeId::of::<T>()).cloned() {
+            id
+        } else {
+            let explicit_bundle_id = self.register_info::<T>(components, storages);
+            // SAFETY: `explicit_bundle_id` is valid and defined above
+            let contributed_components = unsafe {
+                self.get_unchecked(explicit_bundle_id)
+                    .contributed_components()
+                    .to_vec()
+            };
+            let id = self.init_dynamic_info(components, &contributed_components);
```

Comment:
This `to_vec` exists "just" for lifetime reasons. It is sound to pass the original vec into `init_dynamic_info`. Bundle init is performance critical enough to justify using some extra unsafe code here:
```suggestion
            let id = unsafe {
                let (ptr, len) = {
                    // SAFETY: `explicit_bundle_id` is valid and defined above
                    let contributed = self
                        .get_unchecked(explicit_bundle_id)
                        .contributed_components();
                    (contributed.as_ptr(), contributed.len())
                };
                // SAFETY: this is sound because the contributed_components Vec for explicit_bundle_id will not be accessed mutably as
                // part of init_dynamic_info. No mutable references will be created and the allocation will remain valid.
                self.init_dynamic_info(components, std::slice::from_raw_parts(ptr, len))
            };
```